File: String Permutations - GFG/string-permutations.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends



class Solution{
    public:
    //Complete this function
    void findAllPermutations(string s, string permut, vector<string> &ans) {
        if(s.length()==0) {
            ans.push_back(permut);
            return;
        }
        for(int i=0; i<s.length(); i++) {
            permut+=s[i];
            findAllPermutations((s.substr(0, i)+s.substr(i+1)), permut, ans);
            permut.pop_back();
        }        
    }
    vector<string> permutation(string &s)
    {
        vector<string> ans;
        findAllPermutations(s, "", ans);
        sort(ans.begin(), ans.end());
        return ans;
    }
};




//{ Driver Code Starts.

int main()
{
	int T;
	cin>>T;
	while(T--)
	{
		string S;
		cin>>S;
		Solution ob;
		vector<string> vec = ob.permutation(S);
		for(string s : vec){
		    cout<<s<<" ";
		}
		cout<<endl;
	
	}
	return 0;
}
// } Driver Code Ends

File: String Permutations - GFG/string-permutations.py
#User function Template for python3

class Solution:
    def permutation(self,S):
        def help(idx,n,temp,res):
            if idx>=n:
                res.append("".join(temp)) 
                
            for i in range(idx,n):
                temp[i],temp[idx]=temp[idx],temp[i] 
                help(idx+1,n,temp,res)
                temp[i],temp[idx]=temp[idx],temp[i] 
        res=[]
        temp=[x for x in S] 
        # print(temp)
        n=len(S)
        help(0,n,temp,res)
        res.sort()
        # print(res)
        return res


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__=="__main__":
    T=int(input())
    while(T>0):
        s=input()
        ob=Solution()
        ans=ob.permutation(s)
        for i in ans:
            print(i,end=" ")
        print()
        
        T-=1
# } Driver Code Ends